luigi/contrib/external_program.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import os
+import signal
+import subprocess
+import sys
+import tempfile
+
+import luigi
+
+logger = logging.getLogger('luigi-interface')
+
+
+class ExternalProgramTask(luigi.Task):
+    """
+    Template task for running an external program in a subprocess
+
+    The program is run using :py:class:`subprocess.Popen`, with ``args`` passed
+    as a list, generated by :py:meth:`program_args` (where the first element should
+    be the executable). See :py:class:`subprocess.Popen` for details.
+
+    Your must override :py:meth:`program_args` to specify the arguments you want,
+    and you can optionally override :py:meth:`program_environment` if you want to
+    control the environment variables (see :py:class:`ExternalPythonProgramTask`
+    for an example).
+    """
+
+    def program_args(self):
+        """
+        Subclass this method to map your task parameters to the program arguments
+
+        :return: list to pass as ``args`` to :py:class:`subprocess.Popen`
+        """
+        raise NotImplementedError
+
+    def program_environment(self):
+        """
+        Subclass this method to control environment variables for the program
+
+        :return: dict mapping environment variable names to values
+        """
+        env = os.environ.copy()
+        return env
+
+    def _clean_output_file(self, file):
+        file.seek(0)
```

Comment:
it's just a globally defined variable that refers to the `file` class, so one of those things that you *can* override, but generally shouldn't :)